✨ feature[#2] read one product api

sohyun119 opened the pull request:
## ✨ 변경 타입
* [X] 신규 기능 추가/수정
* [ ] 버그 수정
* [X] 리팩토링
* [X] 설정
* [ ] 비기능 (주석 등 기능에 영향을 주지 않음)

## ✅ 체크리스트

* [X] 코드 작성 가이드라인을 준수했는가?
* [X] 변경 사항에 대한 테스트를 완료했는가?
* [ ] 문서 변경이 필요한 경우, 해당 내용을 반영했는가?

## 🚀 변경 내용
* 상품 단건 조회 api 구현
* http client test code 작성
* mapper 빈 주입하지 않고, mapper 안의 메서드 static으로 변경

sohyun119 committed:
🩹 fix [#2] : repository 수정

sohyun119 committed:
✨ feat [#2] : 상품 단건 조회 Controller 구현

sohyun119 committed:
✨ feat [#2] : 상품 단건 조회 ResponseDto 생성 및 구현

sohyun119 committed:
✨ feat [#2] : 상품 예외처리 클래스 추가

sohyun119 committed:
✨ feat [#2] : Mapper 클래스에 toReadResponse 추가

sohyun119 committed:
✨ feat [#2] : 상품 단건 조회 repository 구현

sohyun119 committed:
✨ feat [#2] : 상품 단건 조회 Service 구현

sohyun119 committed:
🩹 fix [#2] : 상품 생성 Response Dto 수정

sohyun119 committed:
🩹 fix [#2] : Component 어노테이션 삭제 및 Static으로 선언
- Service 에서 Mapper를 final로 선언하지 않고 바로 사용

sohyun119 committed:
🍻 test [#2] : http client test code 작성
- 상품 생성 test
- 상품 단건 조회 test

etuhcarap commented on src/main/java/com/sparta/limited/product_service/application/mapper/ProductMapper.java:
```diff
@@ -2,19 +2,23 @@
 import com.sparta.limited.product_service.domain.model.Product;
-import org.springframework.stereotype.Component;
 
-@Component
```
규성님 PR을 확인 하셨군요 👍 👍

sohyun119 replied:
넹 ㅎㅎ 바로 확인하고 가져다 썼습니다!!

etuhcarap commented on src/main/java/com/sparta/limited/product_service/application/service/ProductService.java:
```diff
@@ -16,16 +19,22 @@
+        return ProductMapper.toCreateResponse(product);
     }
 
+    // 상품 단건 조회
```
운영 코드에 주석을 명시하면, 해당 메서드가 어떤 역할을 수행하는지 명시적으로 표현할 수 있지만
반대로 코드가 발전하고 변경됨에 따라 코드에 대한 설명을 포함하는 주석도 같이 관리가 필요해요
따라서, 코드에 반드시 주석이 필요한가 고민해볼 필요가 있어요!
만약 코드의 동작에 대해 추가적인 문서나 설명을 포함하고 싶다면 Test Case를 구성하고, TC를 코드 설명 문서로 대체해볼 수 있어요!

JUnit의 @DisplayName과 같은 어노테이션을 이용해서 동작을 자세히 표현해 볼 수 있으니 이번 기회에 Test Case를 작성해보셔도 좋을 것 같아요!

sohyun119 replied:
네!! 내일 Test Case 작성 바로 해보겠습니다 ㅎㅎㅎ

etuhcarap commented on src/test/resources/http/product-test.http:
```diff
@@ -0,0 +1,13 @@
+
+### 상품 단건 조회
+GET http://localhost:8080/api/v1/products/fcb2d1cf-76e8-4b88-a5a4-5d76571326c5
+Content-Type: application/json
```
빠르게 실행하고 확인할 수 있는 API TC 👍

etuhcarap commented on src/main/java/com/sparta/limited/product_service/infrastructure/repository/ProductRepositoryImpl.java:
```diff
@@ -17,4 +19,9 @@ public void save(Product product) {
 
+    @Override
+    public Optional<Product> findByIdAndDeletedAtIsNull(UUID productId) {
+        return jpaProductRepository.findByIdAndDeletedAtIsNull(productId);
```
매번 쿼리에 삭제되지 않은 상품을 조회하기 위해 deletedAt IS NULL 과 같은 조건을 명시적으로 포함하는 경우, 개발자가 누락하거나 코드 중복이 발생할 수 있어요!

Hibernate에서는 이런 반복되는 코드를 줄이고 일관된 형태로 Entity에 대한 조회 조건을 명시할 수 있는 @SQLRestriction 어노테이션을 제공해요!

@SQLRestriction("deleted_at IS Null")과 같이 어노테이션을 구성하고 Product Entity의 상단에 Global 정책으로 적용해보는건 어떨까요?

sohyun119 replied:
감사합니다! 바로 반영완료했습니다 ㅎㅎ

AnJungGeon replied:
참고하도록 하겠습니다!

etuhcarap reviewed (commented):
주말 늦은 시간까지 화이팅 하시는 모습 보기 좋아요!

참고하실만한 내용을 리뷰로 남겨 보았어요!

다음 PR도 화이팅!

sohyun119 committed:
🩹 fix [#2] : deleted_at IS NULL 조건반영 방법 수정
- @SQLRestriction("deleted_at IS Null") 어노테이션 추가
- findByIdAndDeletedAtIsNULL -> findById 로 변경

sohyun119 committed:
🩹 fix [#2] : 상품 예외처리 메세지 추가

sohyun119 committed:
🩹 fix [#2] : 공통모듈 버전 업데이트

mummumm reviewed (approved):
고생하셨습니다 !! 200점 드립니당👍 👍

chokyuseong00 commented on src/main/java/com/sparta/limited/product_service/domain/model/Product.java:
```diff
@@ -13,11 +13,13 @@
 @Getter
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
 @Table(name = "p_product")
+@SQLRestriction("deleted_at IS Null")
```
@SQLRestriction을 사용하여 관리하시는 부분 좋은 것 같아요! 저도 적용해 보겠습니다!

sohyun119 committed:
🩹 fix [#2] : 공통모듈 버전 업데이트

chokyuseong00 reviewed (approved):
고생하셨습니다!

AnJungGeon reviewed (approved):
수고하셨습니다!
